### ISSUE TITLE
Unable to use 0.4.* BIN compiler with latest package

### ISSUE THREAD

**Issue by blitz-1306** [ethereum/solc-js#602]
## Preface
This is a replication of the issue from https://github.com/ConsenSys/solc-typed-ast/pull/92.

## Setup
- **package.json**
    ```json
    {
        "name": "test-server",
        "version": "1.0.0",
        "description": "",
        "main": "index.js",
        "author": "",
        "license": "UNLICENSED",
        "dependencies": {
            "solc-0811": "npm:solc@0.8.11",
            "solc-0426": "npm:solc@0.4.26",
            "axios": "^0.26.0",
            "fs-extra": "^10.0.0"
        }
    }
    ```
- **index.js**
    ```js
    const axios = require("axios");
    const fse = require("fs-extra");
    const path = require("path");
    const { argv } = require("process");
    const solc0426 = require("solc-0426");
    const solc0811 = require("solc-0811");
    const stream = require('stream');
    const util = require('util');

    const CACHE_DIR = path.join(__dirname, ".compiler_cache");;
    const BINARIES_URL = "https://binaries.soliditylang.org";

    async function getCompilerList(arch) {
        const cachedListPath = path.join(CACHE_DIR, arch, "list.json");

        if (fse.existsSync(cachedListPath)) {
            return fse.readJSONSync(cachedListPath);
        }

        const response = await axios.get(`${BINARIES_URL}/${arch}/list.json`);
        const metaData = response.data;

        fse.ensureDirSync(path.join(CACHE_DIR, arch));
        fse.writeJSONSync(cachedListPath, metaData);

        return metaData;
    }

    async function getCompilerForVersion(arch, version) {
        const list = await getCompilerList(arch);
        const compilerFileName = list.releases[version];

        if (compilerFileName === undefined) {
            return undefined;
        }

        const compilerLocalPath = path.join(CACHE_DIR, arch, compilerFileName);

        if (!fse.existsSync(compilerLocalPath)) {
            const response = await axios({
                method: "GET",
                url: `${BINARIES_URL}/${arch}/${compilerFileName}`,
                responseType: "stream"
            });

            const target = fse.createWriteStream(compilerLocalPath, { mode: 0o555 });
            const pipeline = util.promisify(stream.pipeline);

            await pipeline(response.data, target);
        }

        return compilerLocalPath;
    }

    (async () => {
        const [arch, version, mode] = argv.slice(2);

        console.log(`ARCH: ${arch}`);
        console.log(`VERSION: ${version}`);
        console.log(`MODE: ${mode}`);
        console.log();

        const compilerFileName = await getCompilerForVersion(arch, version);
        const compilerModule = require(compilerFileName);

        if (mode === "legacy") {
            const wrappedCompilerModule = solc0426.setupMethods(compilerModule);
            // const wrappedCompilerModule = solc0811.setupMethods(compilerModule);

            const input = {
                "language": "Solidity",
                "sources": {
                    "sample.sol": "contract Test{}"
                },
                "settings": {
                    "outputSelection": {
                        "*": {
                            "*": [
                                "*"
                            ],
                            "": [
                                "*"
                            ]
                        }
                    }
                }
            }

            const output = wrappedCompilerModule.compile(input);
            // const output = wrappedCompilerModule.compile(JSON.stringify(input));
        
            console.log(output);
        } else {
            const wrappedCompilerModule = solc0811.setupMethods(compilerModule);

            const input = {
                "language": "Solidity",
                "sources": {
                    "sample.sol": {
                        "content": "contract Test{}"
                    } 
                },
                "settings": {
                    "outputSelection": {
                        "*": {
                            "*": [
                                "*"
                            ],
                            "": [
                                "*"
                            ]
                        }
                    }
                }
            }

            const output = wrappedCompilerModule.compile(JSON.stringify(input));
        
            console.log(JSON.parse(output));
        }
    })();

    ```
- **commands**
    ```bash
    npm install
    node index.js bin 0.4.26 legacy
    ```

## Details
This scenario is originally working but as I undestand, SolcJS should make a flow of interaction with compiler to be universal.

If we uncomment lines 68 and 90, then comment lines **67** and **89**, we will get

```
{"errors":[{"component":"general","formattedMessage":"Source input is not a JSON object.","message":"Source input is not a JSON object.","severity":"error","type":"JSONError"}]}
```

If we try to uncomment **89** and comment **90**, then we will get

```
{"errors":[{"component":"general","formattedMessage":"* Line 1, Column 1\n  Syntax error: value, object or array expected.\n* Line 1, Column 1\n  A valid JSON document must be either an array or an object value.\n","message":"* Line 1, Column 1\n  Syntax error: value, object or array expected.\n* Line 1, Column 1\n  A valid JSON document must be either an array or an object value.\n","severity":"error","type":"JSONError"}]}
```

The original expectation was that we are able to use side-loaded compiler as on lines **94** and **117**:
```js
const wrappedCompilerModule = solc0811.setupMethods(compilerModule);
// const input = ...
const output = wrappedCompilerModule.compile(JSON.stringify(input));
```

Can somebody take a look at this? Thanks in advance.

**Comment by axic** [comment]
Sorry, I think this is hard to follow without line numbers when you refer to line numbers.

```
        const output = wrappedCompilerModule.compile(input);
        // const output = wrappedCompilerModule.compile(JSON.stringify(input));
```

However, it always expects stringified JSON as an input.

**Comment by cameel** [comment]
Thanks for reporting!

So the problem is that in the legacy case you're putting the sources directly as the value instead of under `content`:
```json
"sources": {"sample.sol": "contract Test{}"},
```
I don't think 0.4.26 actually supported that. For some reason `solc-js` does not reject it but native `solc --standard-json` does:
```json
{"errors":[{"component":"general","formattedMessage":"Source input is not a JSON object.","message":"Source input is not a JSON object.","severity":"error","type":"JSONError"}]}
```

I checked also 0.4.11 which introduced `--standard-json` and it does not seem like it supported that either. Where are you getting this structure from?

**Comment by axic** [comment]
@cameel using `compile` in latest solc-js equals to the old `compilerStandardWrapper` which tries to smooth out these differences and make the translation. I think it should work as we have tests for 0.4.26 iirc.

**Comment by cameel** [comment]
But did we ever support this structure without `content`? The old docs do not mention it. Or is it for compatibility with pre-Standard JSON input?

In any case, we no longer support it so the question is if it's a regression or if it was dropped on purpose.

**Comment by axic** [comment]
I think your example is not "standard json", but the legacy json input.
